lib/model/model.go

```diff
@@ -1323,6 +1325,32 @@ func (m *Model) Request(deviceID protocol.DeviceID, folder, name string, offset
 		// The folder might be already unpaused in the config, but not yet
 		// in the model.
 		l.Debugf("Request from %s for file %s in unstarted folder %q", deviceID, name, folder)
+		res.Err = protocol.ErrInvalid
+		conn.Response(res)
+		return
+	}
+
+	if err := m.requestPreparation(folderCfg.Filesystem(), folderIgnores, deviceID, folder, name, size, offset, fromTemporary); err != nil {
+		res.Err = err
+		conn.Response(res)
+		return
+	}
+
+	// ToDo: Request limiting (acquire)
+
+	go m.request(conn, res, folderCfg, deviceID, folder, name, size, offset, hash, weakHash, fromTemporary)
+}
+
+func (m *Model) requestPreparation(folderFs fs.Filesystem, folderIgnores *ignore.Matcher, deviceID protocol.DeviceID, folder, name string, size int32, offset int64, fromTemporary bool) error {
+	if offset < 0 {
+		return protocol.ErrInvalid
+	}
+
+	if cfg, ok := m.cfg.Folder(folder); !ok || !cfg.SharedWith(deviceID) {
```

Comment:
The calling method already has a folderCfg that it could give us, instead of us going digging in the config again.